### ISSUE TITLE
unresolved imports when using rust-musl-builder with cargo-chef

### ISSUE THREAD

**Issue by csarn** [LukeMathWalker/cargo-chef#43]
Hi, I'm trying to follow along your excellent upcoming book "Zero to Production", and decided to actually build smaller docker containers than you outlined in chapter 5.
You referenced rust-musl-builder, which I was aware of (and used) before.

It turns out, cargo-chef does not play well with rust-musl-builder.
TL;DR:
- cargo-chef + official rust image works
- rust-musl-builder without cargo-chef works
- Putting both together gives me "unresolved imports" at the build stage

I don't know whether the problem comes from cargo-chef or rust-musl-builder, so I decided to create issues in both projects.
The corresponding issue in rust-musl-builder is https://github.com/emk/rust-musl-builder/issues/115
And for completeness, here is the copy&pasted content of that:

---------------------------

**What did you try to do?**

I have this repository: https://git.christophsarnowski.com/cs/zero2prod
I want to build a Docker image using this rust-musl-builder for static linking, and I have a multi-stage Dockerfile to take advantage of docker layer caching (using cargo-chef).
The dockerfile is this:
https://git.christophsarnowski.com/cs/zero2prod/src/commit/e8fd292295755e5ca037c6814f685c6692f6705c/Dockerfile-alpine

**What happened?**

The 'docker build --tag latest -f Dockerfile-alpine .' fails with the following error at the the 'builder' stage:

```
   Compiling zero2prod v0.1.0 (/home/rust/src/app)
error[E0432]: unresolved imports `zero2prod::configuration`, `zero2prod::startup`, `zero2prod::telemetry`
 --> src/main.rs:6:5
  |
6 |     configuration::get_configuration,
  |     ^^^^^^^^^^^^^ could not find `configuration` in `zero2prod`
7 |     startup::run,
  |     ^^^^^^^ could not find `startup` in `zero2prod`
8 |     telemetry::{get_subscriber, init_subscriber},
  |     ^^^^^^^^^ could not find `telemetry` in `zero2prod`

error: aborting due to previous error

For more information about this error, try `rustc --explain E0432`.
error: could not compile `zero2prod`
```
The "missing imports" do actually exist in the source tree.
This happens locally and in a drone CI environment, and with an empty cache (all layers that should be cached did not exist before this failure. Repeated tries to build the image do use the cache, but fail at the same step with the same message.)

**What did you hope to happen?**

Successful build of the docker image.
In contrast, the other Dockerfile:
https://git.christophsarnowski.com/cs/zero2prod/src/commit/e8fd292295755e5ca037c6814f685c6692f6705c/Dockerfile-debian

with almost exactly the same commands, but based on the official rust image, succeeds in building the image,
as does a local build outside docker.

**Does `./test-image` work?**

I cancelled it about 10 minutes in, it did not show any failures that far. I can report back with a full run.
Also more importantly, building the same application with rust-musl-builder in less stages without cargo-chef worked fine.
It is sufficient to remove the "planner" and "cacher" stages and the "COPY --from=cacher... " lines.

**Comment by mieubrisse** [comment]
Adding a +1 - I'm struggling with the same issue

**Comment by Disco-Dave** [comment]
Hello all, hopefully this is helpful. I am also following along with the book and attempted to utilize cargo-chef with the rust-musl-builder. The following Dockerfile is what I came up with, and is currently working for me.

```Dockerfile
FROM ekidd/rust-musl-builder:1.50.0 AS builder-chef
USER root
ENV CARGO_HOME /usr/var/cargo
RUN cargo install --version 0.1.20 cargo-chef

FROM builder-chef AS planner
WORKDIR /app
COPY --chown=rust:rust . .
RUN cargo chef prepare --recipe-path recipe.json

FROM builder-chef AS cacher
WORKDIR /app
COPY --from=planner --chown=rust:rust /app/recipe.json recipe.json
RUN cargo chef cook --release --target x86_64-unknown-linux-musl --recipe-path recipe.json

FROM ekidd/rust-musl-builder:1.50.0 AS builder
ENV CARGO_HOME /usr/var/cargo
WORKDIR /app
ENV SQLX_OFFLINE true
COPY --chown=rust:rust . .
COPY --from=cacher --chown=rust:rust /app/target target
COPY --from=cacher --chown=rust:rust /usr/var/cargo /usr/var/cargo
RUN cargo build --release --target x86_64-unknown-linux-musl && strip /app/target/x86_64-unknown-linux-musl/release/zero2prod

FROM alpine:3.13 AS runtime
RUN addgroup -S zero2prod && adduser -S zero2prod -G zero2prod
COPY --from=builder /app/target/x86_64-unknown-linux-musl/release/zero2prod  /usr/local/bin/
COPY configuration configuration
ENV APP_ENVIRONMENT production
USER zero2prod
CMD ["/usr/local/bin/zero2prod"]
```

**Comment by michelebariani** [comment]
I also stumbled on this issue. Unluckily for me, the "workaround" by @Disco-Dave is not working on my system/code.

**Comment by LukeMathWalker** [comment]
Can you share your Dockerfile @michelebariani?

**Comment by michelebariani** [comment]
TL;DR: I found my issue to be a different one, the "workaround" by @Disco-Dave is working well also on my system/code. Thanks both @LukeMathWalker and @Disco-Dave

Longer version:

While cleaning up my Dockerfile so to post it here, I had it working at some point...
I am not entirely sure what made the "unresolved imports" go away, there was surely at least one classic "error between keyboard and chair" somewhere, despite all the cache cleanings and rebuilds). Yet I found that behaviour changes depending on the absolute path of the Cargo.lock file within the container instance (maybe it is not even directly related to cargo-chef).
When Cargo.lock is at /Cargo.lock I get trouble, as of now e.g. I have

```
Step 11/22 : RUN cargo chef cook --release --target x86_64-unknown-linux-musl --recipe-path recipe.json
 ---> Running in 66fabfb31d55
error: failed to parse lock file at: /Cargo.lock

Caused by:
  package `cfg-if` is specified twice in the lockfile
thread 'main' panicked at 'Exited with status code: 101', /usr/local/cargo/registry/src/github.com-1ecc6299db9ec823/cargo-chef-0.1.24/src/recipe.rs:142:27
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

and that of course points to the further general topic of duplicates (I checked the file, it is actually present two times with two different versions). Yet curiously enough, everything goes smoothly if the very same file is e.g. at /app/Cargo.lock

**Comment by LukeMathWalker** [comment]
I have updated the README (#90) to recommend @Disco-Dave's approach.
I'll close this issue, as the approach seems to work, but feel free to re-open if there are new edge cases that need to be considered.

**Comment by dzmitry-lahoda** [comment]
I get exactly same error in remote ci docker, while not getting it locally, so recipe locally contains cfg-if 2 times
```
package `cfg-if` is specified twice in the lockfile
```

**Comment by dzmitry-lahoda** [comment]
my toml not in app, so it is in subfolder of git
